### PR TITLE
Fix toolstate for rust-lang/rust#85546

### PR DESCRIPTION
**Pull request by hyd-dev**
cc rust-lang/rust#85780

### CODE REVIEW

**Comment by hyd-dev** [comment]
This is probably out of sync with rust-lang/rust#85546 now because I changed many functions of `Machine` in the latest commit... I'll update this later.

File: src/shims/panic.rs [hunk 1]
```diff
@@ -40,15 +40,11 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
...
         &mut self,
         args: &[OpTy<'tcx, Tag>],
-        unwind: Option<mir::BasicBlock>,
+        unwind: StackPopUnwind,
     ) -> InterpResult<'tcx> {
         let this = self.eval_context_mut();
 
         trace!("miri_start_panic: {:?}", this.frame().instance);
-        // Make sure we only start unwinding when this matches our panic strategy.
-        if this.tcx.sess.panic_strategy() != PanicStrategy::Unwind {
```
Comment (RalfJung) [inline src/shims/panic.rs:49]: Why did you remove this?

File: src/shims/panic.rs [hunk 2]
```diff
@@ -58,7 +54,13 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         thread.panic_payload = Some(payload);
 
         // Jump to the unwind block to begin unwinding.
-        this.unwind_to_block(unwind);
+        this.unwind_to_block(match unwind {
```
Comment (RalfJung) [inline src/shims/panic.rs:57]: This looks like `unwind_to_block` should really take a `StackPopUnwind`.

Comment (hyd-dev) [reply src/shims/panic.rs:57]: That's a great idea, and that means this `throw_ub_format!` should be moved inside `unwind_to_block()`, and `pop_stack_frame()` calls `unwind_to_block()`, which means it also does not need its own `throw_ub_format!` anymore. However, that also means we can't do https://github.com/rust-lang/rust/pull/85546#discussion_r638036854 anymore, unless we want to duplicate this check before popping the frame, or pushing the frame back before `throw_ub_format!`.

Comment (RalfJung) [reply src/shims/panic.rs:57]: Ah, good point... maybe we leave improving the span to the future? FWIW, returning from a function when the return block is `None` has the exact same problem; that error is currently checked in `return_to_block`.

File: src/shims/panic.rs (hunk 1 above)
Comment (hyd-dev) [reply src/shims/panic.rs:49]: `unwind_to_block()` will check this now: https://github.com/rust-lang/rust/blob/1969c2e312303faa458cf19dad9783165e3c72c8/compiler/rustc_middle/src/ty/layout.rs#L2588
